### ISSUE TITLE
"--hls-use-mpegts --no-part" should create files with ".ts" extension

### ISSUE THREAD

**Issue by Vangelis66** [rg3/youtube-dl#11105]
OS: Windows Vista SP2 x86, latest OS updates from Microsoft.
Using the latest standalone "youtube-dl.exe" downloaded from Github.

`youtube-dl --version => 2016.11.02`

Also using FFmpeg.exe 3.0 (x86) downloaded from the Zeranoe repo.

The first channel of the Greek National TV broadcaster uses Youtube to host its live stream; 
[ERT1 LIVE](http://webtv.ert.gr/ert1/) => [https://www.youtube.com/watch?v=gnnO4TeJG-4](https://www.youtube.com/watch?v=gnnO4TeJG-4) 
The actual youtube URL changes frequently... 
`--hls-prefer-native` doesn't seem to be able to handle this stream, 
so ffmpeg MUST be used. When issuing: 

`youtube-dl -f 94 --no-part "https://www.youtube.com/watch?v=gnnO4TeJG-4"` 

the stream download is delegated to FFmpeg and (correctly) youtube-dl creates 
a file with the ".mp4" extension. When I `CTRL+C`, the end file is correctly 
identified by MediaInfo to be an MP4 file.
On the contrary, when I issue: 

`youtube-dl -f 94 --hls-use-mpegts --no-part "https://www.youtube.com/watch?v=gnnO4TeJG-4"`

youtube-dl (again) generates a file with file extension ".mp4", but 
the `--hls-use-mpegts` switch creates an MPEG-TS container which has 
(by default) a ".ts" file extension (... at least on Windows)! 
`CTRL+C`ing results in an end file which is, in essence, a ".ts" file, 
as reported by MediaInfo, but with an erroneous "mp4" extension. 
I have to manually rename the file back to the correct ".ts" extension. 

 I am of the opinion this is a bugged behaviour on the part of the 
application and that youtube-dl should be smart enough to attribute 
the correct file extention (.ts) when `--hls-use-mpegts` is employed...

Many thanks to all the devs' team for their on-going efforts!

**Comment by Vangelis66** [comment]
@jaimeMF 

Thanks for reading my report; though, if you ask me, 
the "**bug**" label would have been more appropriate... 

I am not "**request**ing" some extra (new) feature of youtube-dl 
(the case on most issues [here](https://github.com/rg3/youtube-dl/labels/request)), rather pointing out 
that an existing feature be ironed out! 
Anyhow, keep up the good work!

**Comment by SteveJobzniak** [comment]
@jaimeMF @Vangelis66 I agree that it should do it by default, because it's silly to have to override the format.

But @Vangelis66, I think all modules default to `"%(title)s-%(id)s.%(ext)s"`, so instead of just sending --hls-use-mpegts, you should instead send `--hls-use-mpegts --output "%(title)s-%(id)s.ts"`.

But I agree that this should be done by default... if possible.

**Comment by diamond12** [comment]
@SteveJobzniak Or maybe it is easier to replace ext with .ts if ffmpeg is executed with the `-f mpegts` parameter.

**Comment by SteveJobzniak** [comment]
@diamond12 Not sure if that would work. The "external downloader" script that launches ffmpeg says `ffmpeg -f mpegts` every time you use `youtube-dl --hls-use-mpegts`, even if the actual file is not mpegts. See: https://github.com/rg3/youtube-dl/blob/master/youtube_dl/downloader/external.py#L264

And https://github.com/rg3/youtube-dl/blob/master/youtube_dl/downloader/hls.py is another HLS downloader which would need fixing too.

I propose a possible solution: If the youtube-dl extractor detects a .m3u8 file AND `--hls-use-mpegts` is passed as a parameter, then rewrite `%(ext)s` to "ts". That means that anytime "ext" is used in the output filename pattern, it would correctly use ".ts" - IF - the stream is HLS and is being saved as mpegts.

Either way, this is a problem in youtube-dl. Its "automatic filename" feature shouldn't apply a totally incorrect file extension like this.

**Comment by diamond12** [comment]
@SteveJobzniak I agree it's a bug and I don't know why it is marked as a feature request.

The source that you pointed me to is right after `if protocol in ('m3u8', 'm3u8_native')` so it should be fine.  Besides, `-f mpegts` will always produce a .ts.  But yes, your proposed solution is a good option.

**Comment by SteveJobzniak** [comment]
@diamond12 Oh yeah you are right! So they _do_ check that it's a HLS stream before applying --hls-use-mpegts. Nice find! That should mean it's possible to always run with --hls-use-mpegts even if the URL isn't HLS.

All that's missing now is applying the proper output extension (.ts) when saving a HLS.

I think changing the meaning of `ext` to `ts` at the moment when the output filename pattern is parsed (as described above) makes the most sense. That way it fixes the `ext` portion of the filename everywhere in the code that tries to look at / use the extension, and even in any user-provided filename patterns on the command-line (if the user uses `%(ext)s`.

I am very overworked but took about a minute to look at the source. It seems the function `process_video_result` is one possible (bad?) injection point:

https://github.com/rg3/youtube-dl/blob/master/youtube_dl/YoutubeDL.py#L1363

It loops through all possible video sources and fills in their 'ext' value. And then it also runs `determine_protocol ` (which sets the value to `m3u8` or `m3u8_native`?).

The `determine_protocol` code is in utils:

https://github.com/rg3/youtube-dl/blob/master/youtube_dl/utils.py#L2323

It seems like all protocol/URL-determining code is done without connecting to the server to check any mimetype. So @jaimeMF What happens in YouTube-DL (in general), if a URL ends in something like "/streamplaylist.php?channel=whatever" and streamplaylist.php returns the M3U8 mimetype + playlist-data? Does YouTube-DL detect that when it connects to the server and quickly fix itself to get ready to download as HLS? Or does it decide to download everything with a ".php" extension and "unknown video format"?

Here are the two injection points I can see:

A. Best method, if YouTube-DL is capable of determining mime-type and fixing its detection of the "format" when it actually connects to a server (but it doesn't seem like YouTube-DL can do that):

- Connect to the stream URL and determine the format.
- If the mimetype is m3u8 and the user has provided `--hls-use-mpegts`, fix the output `ext` to `ts`.

B. Alternative method, which seems like **the one to use**, because it seems like YouTube-DL guesses protocol+ext without connecting to the server, and then just passes those parameters to the downloader (such as ffmpeg):

- Modify the extension selector in `process_video_result`.
- If `--hls-use-mpegts` was provided as an option, and the format is detected as `m3u8` or `m3u8_native`, then rewrite `format['ext'] = 'ts`.

**Comment by SteveJobzniak** [comment]
@jaimeMF I am extremely overworked, but took a minute to look at the source to get things rolling. It seems the function `process_video_result` is the correct injection point:

https://github.com/rg3/youtube-dl/blob/master/youtube_dl/YoutubeDL.py#L1363

It loops through all possible video sources and fills in their 'ext' value. And then it also runs `determine_protocol ` (which sets the value to `m3u8` or `m3u8_native`?).

The `determine_protocol` code is in utils:

https://github.com/rg3/youtube-dl/blob/master/youtube_dl/utils.py#L2323

It seems like all of this extension/protocol/URL-determining code is done _without_ connecting to the server to check any mimetype. It looks like i's up to that website's extractor (for example https://github.com/rg3/youtube-dl/blob/master/youtube_dl/extractor/animeondemand.py) to put in a `protocol` value for use by YouTube-DL (https://github.com/rg3/youtube-dl/blob/master/youtube_dl/YoutubeDL.py#L1359). So YouTube-DL doesn't connect any further to the URL it's given to the extractor. It just passes everything to the desired downloader method.

Modifying the extension selector in YoutubeDL.py's `process_video_result` is the correct location for the bugfix, from what I can see:

- First, let it run the existing code that does server-side format determination if none was provided by the extractor (`if format.get('format') is None:`), so that we know what file format the URL is providing.
- Next, let it run the existing code that does extension determination if none was provided by the extractor (`if format.get('ext') is None:`), so that we know what extension to use for the output file in normal situations.
- Next, let it run the existing code that does protocol determination if none was provided by the extractor (`if format.get('protocol') is None:`), so that we see when it's a `m3u8` or `m3u8_native` protocol. This just looks at the extension/URL so it's quite basic. So in case the website uses something like a .php script as I theorized above, then it'd be up to the extractor to actually provide a proper `protocol` value to YouTube-DL.
- **After all of that, we know the format, extension and protocol**, so then just add the following code: If `--hls-use-mpegts` was provided as an option to YouTube-DL (maybe using the same `if self.params.get('hls_use_mpegts', False) or tmpfilename == '-':` check that external.py uses? https://github.com/rg3/youtube-dl/blob/master/youtube_dl/downloader/external.py#L264), and the protocol is detected as `m3u8` or `m3u8_native`, then rewrite `format['ext'] = 'ts`.

That would fix the bug. Please comment to let us know your thoughts. Is there a better method? Did I find the right method? Could you fix this without a pull request?
